### ISSUE TITLE
Resource abstractor and lib issues

### ISSUE THREAD

**Issue by giobart** [oakestra/oakestra#324]
## Short
The latest PR introducing resoruce abstractor in system manager is causing issues with the latest alpha images

## Reproduce the issue
- Move to branch [update-alpha-overrides](https://github.com/oakestra/oakestra/tree/update-alpha-overrides)
- Run a 1doc cluster using the latest overrides for alpha-v0.4.302
```
sudo -E docker compose -f 1-DOC.yaml -f override-alpha-versions.yaml up
```

When attaching a cluster:
```
system_manager               | Traceback (most recent call last):
system_manager               |   File "/usr/local/lib/python3.10/site-packages/eventlet/hubs/hub.py", line 476, in fire_timers
system_manager               |     timer()
system_manager               |   File "/usr/local/lib/python3.10/site-packages/eventlet/hubs/timer.py", line 59, in __call__
system_manager               |     cb(*args, **kw)
system_manager               |   File "/usr/local/lib/python3.10/site-packages/eventlet/greenthread.py", line 221, in main
system_manager               |     result = function(*args, **kwargs)
system_manager               |   File "/usr/local/lib/python3.10/site-packages/socketio/server.py", line 586, in _handle_event_internal
system_manager               |     r = server._trigger_event(data[0], namespace, sid, *data[1:])
system_manager               |   File "/usr/local/lib/python3.10/site-packages/socketio/server.py", line 611, in _trigger_event
system_manager               |     return handler(*args)
system_manager               |   File "/usr/local/lib/python3.10/site-packages/flask_socketio/__init__.py", line 282, in _handler
system_manager               |     return self._handle_event(handler, message, namespace, sid,
system_manager               |   File "/usr/local/lib/python3.10/site-packages/flask_socketio/__init__.py", line 827, in _handle_event
system_manager               |     ret = handler(*args)
system_manager               |   File "//system_manager.py", line 115, in handle_init_client
system_manager               |     cluster = cluster_operations.upsert_cluster(cluster_data)
system_manager               | AttributeError: module 'resource_abstractor_client.cluster_operations' has no attribute 'upsert_cluster'
```

When creating an application:
```
resource_abstractor          | ::ffff:172.18.0.18 - - [06/May/2024 09:06:12] "POST /api/v1/applications/ HTTP/1.1" 200 -
resource_abstractor          | ::ffff:172.18.0.18 - - [06/May/2024 09:06:12] "GET /api/v1/applications?userId=Admin HTTP/1.1" 308 -
resource_abstractor          | ::ffff:172.18.0.18 - - [06/May/2024 09:06:12] "GET /api/v1/applications/?userId=Admin HTTP/1.1" 200 -
system_manager               | 2024-05-06 09:06:12,234 - system_manager - INFO - Traceback (most recent call last):
system_manager               |   File "/usr/local/lib/python3.10/site-packages/eventlet/wsgi.py", line 569, in handle_one_response
system_manager               |     result = self.application(self.environ, start_response)
system_manager               |   File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 2091, in __call__
system_manager               |     return self.wsgi_app(environ, start_response)
system_manager               |   File "/usr/local/lib/python3.10/site-packages/flask_socketio/__init__.py", line 43, in __call__
system_manager               |     return super(_SocketIOMiddleware, self).__call__(environ,
system_manager               |   File "/usr/local/lib/python3.10/site-packages/engineio/middleware.py", line 74, in __call__
system_manager               |     return self.wsgi_app(environ, start_response)
system_manager               |   File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 2076, in wsgi_app
system_manager               |     response = self.handle_exception(e)
system_manager               |   File "/usr/local/lib/python3.10/site-packages/flask_cors/extension.py", line 178, in wrapped_function
system_manager               |     return cors_after_request(app.make_response(f(*args, **kwargs)))
system_manager               |   File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 2073, in wsgi_app
system_manager               |     response = self.full_dispatch_request()
system_manager               |   File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 1519, in full_dispatch_request
system_manager               |     rv = self.handle_user_exception(e)
system_manager               |   File "/usr/local/lib/python3.10/site-packages/flask_cors/extension.py", line 178, in wrapped_function
system_manager               |     return cors_after_request(app.make_response(f(*args, **kwargs)))
system_manager               |   File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 1517, in full_dispatch_request
system_manager               |     rv = self.dispatch_request()
system_manager               |   File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 1503, in dispatch_request
system_manager               |     return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
system_manager               |   File "/usr/local/lib/python3.10/site-packages/flask/views.py", line 84, in view
system_manager               |     return current_app.ensure_sync(self.dispatch_request)(*args, **kwargs)
system_manager               |   File "/usr/local/lib/python3.10/site-packages/flask_restful/__init__.py", line 604, in dispatch_request
system_manager               |     resp = meth(*args, **kwargs)
system_manager               |   File "/usr/local/lib/python3.10/site-packages/webargs/core.py", line 649, in wrapper
system_manager               |     return func(*args, **kwargs)
system_manager               |   File "/usr/local/lib/python3.10/site-packages/flask_smorest/arguments.py", line 82, in wrapper
system_manager               |     return func(*f_args, **f_kwargs)
system_manager               |   File "/usr/local/lib/python3.10/site-packages/flask_smorest/response.py", line 90, in wrapper
system_manager               |     func(*args, **kwargs)
system_manager               |   File "/usr/local/lib/python3.10/site-packages/flask_jwt_extended/view_decorators.py", line 170, in decorator
system_manager               |     return current_app.ensure_sync(fn)(*args, **kwargs)
system_manager               |   File "/blueprints/applications_blueprints.py", line 102, in post
system_manager               |     return json_util.dumps({**result, "_id": {"$oid": str(result["_id"])}})
system_manager               | TypeError: 'list' object is not a mapping
system_manager               | 
system_manager               | 2024-05-06 09:06:12,235 - system_manager - INFO - ::ffff:172.24.146.26 - - [06/May/2024 09:06:12] "POST /api/application/ HTTP/1.1" 500 2848 0.045511
```

When removing an application:
```
resource_abstractor          | [2024-05-06 09:28:25,408] ERROR in app: Exception on /api/v1/applications/66389d84961f55d48c870e3c [DELETE]
resource_abstractor          | Traceback (most recent call last):
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 2073, in wsgi_app
resource_abstractor          |     response = self.full_dispatch_request()
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1519, in full_dispatch_request
resource_abstractor          |     rv = self.handle_user_exception(e)
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1517, in full_dispatch_request
resource_abstractor          |     rv = self.dispatch_request()
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1503, in dispatch_request
resource_abstractor          |     return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/flask/views.py", line 84, in view
resource_abstractor          |     return current_app.ensure_sync(self.dispatch_request)(*args, **kwargs)
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/flask/views.py", line 158, in dispatch_request
resource_abstractor          |     return current_app.ensure_sync(meth)(*args, **kwargs)
resource_abstractor          |   File "/api/v1/apps_blueprint.py", line 42, in delete
resource_abstractor          |     return json.dumps(apps_db.delete_app(appId), default=str)
resource_abstractor          |   File "/db/apps_db.py", line 18, in delete_app
resource_abstractor          |     return db.mongo_apps.find_one_and_delete(filter, return_document=True)
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/pymongo/collection.py", line 3202, in find_one_and_delete
resource_abstractor          |     return self.__find_and_modify(
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/pymongo/collection.py", line 3110, in __find_and_modify
resource_abstractor          |     return self.__database.client._retryable_write(
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/pymongo/mongo_client.py", line 1569, in _retryable_write
resource_abstractor          |     return self._retry_with_session(retryable, func, s, bulk, operation, operation_id)
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/pymongo/mongo_client.py", line 1455, in _retry_with_session
resource_abstractor          |     return self._retry_internal(
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/pymongo/_csot.py", line 108, in csot_wrapper
resource_abstractor          |     return func(self, *args, **kwargs)
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/pymongo/mongo_client.py", line 1490, in _retry_internal
resource_abstractor          |     return _ClientConnectionRetryable(
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/pymongo/mongo_client.py", line 2347, in run
resource_abstractor          |     return self._read() if self._is_read else self._write()
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/pymongo/mongo_client.py", line 2464, in _write
resource_abstractor          |     return self._func(self._session, conn, self._retryable)  # type: ignore
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/pymongo/collection.py", line 3096, in _find_and_modify
resource_abstractor          |     out = self._command(
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/pymongo/collection.py", line 301, in _command
resource_abstractor          |     return conn.command(
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/pymongo/helpers.py", line 327, in inner
resource_abstractor          |     return func(*args, **kwargs)
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/pymongo/pool.py", line 985, in command
resource_abstractor          |     return command(
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/pymongo/network.py", line 212, in command
resource_abstractor          |     helpers._check_command_response(
resource_abstractor          |   File "/usr/local/lib/python3.8/site-packages/pymongo/helpers.py", line 233, in _check_command_response
resource_abstractor          |     raise OperationFailure(errmsg, code, response, max_wire_version)
resource_abstractor          | pymongo.errors.OperationFailure: Cannot specify both new=true and remove=true; 'remove' always returns the deleted document, full error: {'ok': 0.0, 'errmsg': "Cannot specify both new=true and remove=true; 'remove' always returns the deleted document", 'code': 9, 'codeName': 'FailedToParse'}
```

## Solution 
One of the issues is caused by a missing method in resource abstractor library. The rest must be investigated further. 

## Status
Looking for a solution

## Checklist
- [ ] Discussed
- [ ] Solved
- [ ] Tested

**Comment by giobart** [comment]
I re-open the issue but for a different matter. When deploying an application from frontened we observe the following issue

```
resource_abstractor          | ::ffff:172.20.0.18 - - [07/May/2024 13:05:14] "GET /api/v1/jobs/undefined HTTP/1.1" 400 -
```

Here the full log

```
cluster_manager              | 2024-05-07 13:05:12,873 - cluster_manager - INFO - MQTT - Received from worker: 
cluster_manager              | 2024-05-07 13:05:12,874 - cluster_manager - INFO - {'topic': 'nodes/663a2679bd20f011ce76faa7/jobs/resources', 'payload': '{"services":[]}'}
cluster_manager              | 2024-05-07 13:05:12,900 - cluster_manager - INFO - MQTT - Received from worker: 
cluster_manager              | 2024-05-07 13:05:12,900 - cluster_manager - INFO - {'topic': 'nodes/663a2679bd20f011ce76faa7/jobs/resources', 'payload': '{"services":[]}'}
cluster_manager              | 2024-05-07 13:05:13,396 - cluster_manager - INFO - MQTT - Received from worker: 
cluster_manager              | 2024-05-07 13:05:13,396 - cluster_manager - INFO - {'topic': 'nodes/663a2679bd20f011ce76faa7/information', 'payload': '{"id":"","host":"","ip":"","port":"","system_info":null,"cpu":0.3,"free_cores":6,"architecture":"","memory":18.177810195479754,"memory_free_in_MB":6001,"disk_info":null,"network_info":null,"gpu_driver":"-","gpu_usage":0,"gpu_cores":0,"gpu_temp":0,"gpu_mem_used":0,"gpu_tot_mem":0,"technology":null,"Overlay":false,"LogDirectory":"","NetManagerPort":0}'}
cluster_manager              | 2024-05-07 13:05:13,396 - cluster_manager - INFO - MONGODB - update cpu and memory of worker node 663a2679bd20f011ce76faa7 ...
system_manager               | 2024-05-07 13:05:14,010 - system_manager - INFO - ::ffff:172.24.146.18 - - [07/May/2024 13:05:14] "OPTIONS /api/service/undefined/instance HTTP/1.1" 200 362 0.002114
system_manager               | 2024-05-07 13:05:14,011 - system_manager - INFO - ::ffff:172.24.146.18 - - [07/May/2024 13:05:14] "OPTIONS /api/service/undefined/instance HTTP/1.1" 200 362 0.000518
resource_abstractor          | ::ffff:172.20.0.18 - - [07/May/2024 13:05:14] "GET /api/v1/jobs/undefined HTTP/1.1" 400 -
system_manager               | 2024-05-07 13:05:14,092 - system_manager - INFO - ::ffff:172.24.146.18 - - [07/May/2024 13:05:14] "POST /api/service/undefined/instance HTTP/1.1" 200 220 0.013225
resource_abstractor          | ::ffff:172.20.0.18 - - [07/May/2024 13:05:14] "GET /api/v1/jobs/undefined HTTP/1.1" 400 -
system_manager               | 2024-05-07 13:05:14,097 - system_manager - INFO - ::ffff:172.24.146.18 - - [07/May/2024 13:05:14] "POST /api/service/undefined/instance HTTP/1.1" 200 220 0.004428
```


application deletion instead fails for some undetermined reason without stack trace but simply with code 500:
```
system_manager               | 2024-05-07 13:07:55,219 - system_manager - INFO - ::ffff:172.24.146.18 - - [07/May/2024 13:07:55] "OPTIONS /api/service/undefined HTTP/1.1" 200 380 0.001975
resource_abstractor          | ::ffff:172.20.0.18 - - [07/May/2024 13:07:55] "GET /api/v1/applications?userId=Admin HTTP/1.1" 308 -
resource_abstractor          | ::ffff:172.20.0.18 - - [07/May/2024 13:07:55] "GET /api/v1/applications/?userId=Admin HTTP/1.1" 200 -
system_manager               | 2024-05-07 13:07:55,309 - system_manager - INFO - ::ffff:172.24.146.18 - - [07/May/2024 13:07:55] "DELETE /api/service/undefined HTTP/1.1" 500 273 0.022398
```

## The issue

It looks like the frontend does not get app ID but obtains `undefined` instead. 

What is causing it: we have an inconsistency in how we return the _id field in Services.

Before: `"_id\": {\"$oid\": \"663a2a44735e95f205231287\"}`
Now: `"_id\": "663a2a44735e95f205231287`

But for applications we still return: `"_id\": {\"$oid\": \"663a2a44735e95f205231287\"}`. 

I fully support the removal of the nested "oid" field, but either in all cases or none. I would suggest  in this api version to go back to `"_id\": {\"$oid\": \"663a2a44735e95f205231287\"}` and in api v2 we'll remove "oid" all together. what do you think @melkodary?

**Comment by melkodary** [comment]
Thank you for providing all these details.

That's a reasonable approach you explained there. I'll add a fix for the jobs/services too. @giobart

**Comment by Malyuk-A** [comment]
Can this issue be closed because #328 was merged?

**Comment by oakestra-admin** [comment]
🥀 This issue is stale because it has been open 30 days with no activity. Remove stale label or comment or this will be closed in 10 days.
